File: log-parser/app/analysis/medal_rate_ema.py
import math
from typing import Optional
from datetime import datetime


class MedalRateEMA:
    def __init__(self, decay_const: float = 20.0):
        self.decay_const = decay_const
        self.last_timestamp: Optional[datetime] = None
        self.last_total: Optional[int] = None
        self.ema_rate: float = 0.0
        self.offset_total: int = 0

    def add_offset(self, value: int) -> None:
        self.offset_total += value

    def update(self, total: int, timestamp: datetime) -> Optional[int]:
        adjusted_total = total - self.offset_total

        if self.last_timestamp is None:
            self.last_timestamp = timestamp
            self.last_total = adjusted_total
            return None

        dt = (timestamp - self.last_timestamp).total_seconds()
        if dt <= 0:
            return int(self.ema_rate)

        delta = adjusted_total - self.last_total
        rate_instant = (delta / dt) * 60.0  # [medals/min]

        alpha = 1 - math.exp(-dt / self.decay_const)

        self.ema_rate = (1 - alpha) * self.ema_rate + alpha * rate_instant

        self.last_timestamp = timestamp
        self.last_total = adjusted_total

        return int(self.ema_rate)

    def reset(self) -> None:
        self.last_timestamp = None
        self.last_total = None
        self.ema_rate = 0.0
        self.offset_total = 0

File: log-parser/internal/analysis/medal_rate_ema.go
package analysis

import (
	"math"
	"time"
)

type MedalRateEMA struct {
	decayConst    float64
	lastTimestamp *time.Time
	lastTotal     *int64
	emaRate       float64
	offsetTotal   int64
}

func NewMedalRateEMA(decayConst float64) *MedalRateEMA {
	return &MedalRateEMA{decayConst: decayConst}
}

func (e *MedalRateEMA) AddOffset(value int64) {
	e.offsetTotal += value
}

func (e *MedalRateEMA) Update(total int64, ts time.Time) *int64 {
	adjusted := total - e.offsetTotal

	if e.lastTimestamp == nil {
		e.lastTimestamp = &ts
		e.lastTotal = &adjusted
		return nil
	}

	dt := ts.Sub(*e.lastTimestamp).Seconds()
	if dt <= 0 {
		r := int64(e.emaRate)
		return &r
	}

	delta := float64(adjusted - *e.lastTotal)
	instantRate := (delta / dt) * 60.0 // medals/min
	alpha := 1 - math.Exp(-dt/e.decayConst)
	e.emaRate = (1-alpha)*e.emaRate + alpha*instantRate

	e.lastTimestamp = &ts
	e.lastTotal = &adjusted

	r := int64(e.emaRate)
	return &r
}

func (e *MedalRateEMA) Reset() {
	e.lastTimestamp = nil
	e.lastTotal = nil
	e.emaRate = 0
	e.offsetTotal = 0
}
